### PR TITLE
Add `field_property` and `accessor_property` helper macro examples and docs

### PR DESCRIPTION
**Pull request by ginger51011**

### CODE REVIEW

File: trustfall_core/src/interpreter/helpers.rs [hunk 1]
```diff
@@ -8,6 +8,9 @@ use super::basic_adapter::{ContextIterator, ContextOutcomeIterator, VertexIterat
 ///
 /// Takes a property-resolver function and applies it over each of the vertices
 /// in the input context iterator, one at a time.
+///
+/// Often used with the [`field_property!`](crate::field_property) and
+/// [`accessor_property!`](crate::accessor_property)
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:13]:
```suggestion
/// Often used with resolvers from the [`field_property!`](crate::field_property) and
/// [`accessor_property!`](crate::accessor_property) macros.
```

File: trustfall_core/src/interpreter/helpers.rs [hunk 2]
```diff
@@ -63,6 +66,110 @@ pub fn resolve_coercion_with<'vertex, Vertex: Debug + Clone + 'vertex>(
 }
 
 /// Helper for making property resolver functions based on fields.
+///
+/// Generally used with [`resolve_property_with`].
+///
+/// Retrieves a [FieldValue] from a vertex by converting it to the proper type,
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:72]: I'm not a docs expert, do we generally want to put type names in code blocks or is there some reason it's better not to here?
```suggestion
/// Retrieves a [`FieldValue`] from a vertex by converting it to the proper type,
```

File: trustfall_core/src/interpreter/helpers.rs [hunk 3]
```diff
@@ -63,6 +66,110 @@ pub fn resolve_coercion_with<'vertex, Vertex: Debug + Clone + 'vertex>(
...
+/// # Examples
+/// ```
+/// # use std::rc::Rc;
+/// # use trustfall_core::{
+/// #    field_property,
+/// #    ir::FieldValue,
+/// #    interpreter::{
+/// #            helpers::resolve_property_with, basic_adapter::{ContextIterator, ContextOutcomeIterator}
+/// #        }
+/// #    };
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:87]:
```suggestion
/// # use trustfall_core::{
/// #     field_property,
/// #     interpreter::{
/// #         basic_adapter::{ContextIterator, ContextOutcomeIterator},
/// #         helpers::resolve_property_with,
/// #     },
/// #     ir::FieldValue,
/// # };
```
I just ran this through `cargo fmt`.

File: trustfall_core/src/interpreter/helpers.rs [hunk 4]
```diff
@@ -63,6 +66,110 @@ pub fn resolve_coercion_with<'vertex, Vertex: Debug + Clone + 'vertex>(
...
+/// in the example below.
+/// ```
+/// # use std::rc::Rc;
+/// # use trustfall_core::{
+/// #    field_property,
+/// #    ir::FieldValue,
+/// #    interpreter::{
+/// #            helpers::resolve_property_with, basic_adapter::{ContextIterator, ContextOutcomeIterator}
+/// #        }
+/// #    };
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:123]:
```suggestion
/// # use trustfall_core::{
/// #     field_property,
/// #     interpreter::{
/// #         basic_adapter::{ContextIterator, ContextOutcomeIterator},
/// #         helpers::resolve_property_with,
/// #     },
/// #     ir::FieldValue,
/// # };
```

File: trustfall_core/src/interpreter/helpers.rs [hunk 5]
```diff
@@ -63,6 +66,110 @@ pub fn resolve_coercion_with<'vertex, Vertex: Debug + Clone + 'vertex>(
...
+/// # ) -> ContextOutcomeIterator<'static, Vertex, FieldValue> {
+/// #    match (type_name, property_name) {
+/// ("User" | "Bot", "id") => {
+///     resolve_property_with(contexts, field_property!(as_user, id)) // Macro used here
+/// },
+/// #        // ...
+/// #        _ => unreachable!()
+/// #    }
+/// # }
+/// ```
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:169]: Love this, great idea to combine two vertex variants with `as_user` like this!

File: trustfall_core/src/interpreter/helpers.rs [hunk 6]
```diff
@@ -88,6 +195,59 @@ macro_rules! field_property {
...
+///
+/// In principle exactly the same as [`field_property!`](crate::field_property),
+/// but where the property is to be accessed using an accessor function instead
+/// of as a field.
+///
+/// # Examples
+///
+/// In the following example, `name` would be accessed using a field, but the
+/// age is accessed using a function
+///
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:207]:
```suggestion
/// age is accessed using a function:
```

File: trustfall_core/src/interpreter/helpers.rs [hunk 7]
```diff
@@ -88,6 +195,59 @@ macro_rules! field_property {
...
+/// ```rust
+/// # use std::rc::Rc;
+/// # use trustfall_core::{
+/// #    field_property,
+/// #    accessor_property,
+/// #    ir::FieldValue,
+/// #    interpreter::{
+/// #            helpers::resolve_property_with, basic_adapter::{ContextIterator, ContextOutcomeIterator}
+/// #        }
+/// #    };
```
Comment (obi1kenobi) [inline trustfall_core/src/interpreter/helpers.rs:217]:
```suggestion
/// # use trustfall_core::{
/// #     accessor_property,
/// #     field_property,
/// #     interpreter::{
/// #         basic_adapter::{ContextIterator, ContextOutcomeIterator},
/// #         helpers::resolve_property_with,
/// #     },
/// #     ir::FieldValue,
/// # };
```

**Review by obi1kenobi — COMMENTED** [review]
Awesome stuff, as always!

File: trustfall_core/src/interpreter/helpers.rs (hunk 2 above)
Comment (ginger51011) [reply trustfall_core/src/interpreter/helpers.rs:72]: You're right, I though it automatically code-formatted when it was a link to code.

**Comment by obi1kenobi** [comment]
Awesome stuff!

Btw, do you think a macro diagram like this might help users understand the macros better, or would it just be too confusing? https://twitter.com/Lucretiel/status/1625184904500723714

**Pull request merged**